File: Frogger/light.cpp
#include "light.h"

Light::Light() {

	index = 0;
	vsml = VSMathLib::getInstance();

	for (int i = 0; i < MAX_LIGHTS; i++) {
		myLightSources[i].isEnabled = false;
	}
}

Light::~Light() {}

void Light::addLight(LightTypes lightTypes, float *pos, float *spotDir, float spotCutOff, int index) {

	if (index < MAX_LIGHTS) {

		switch (lightTypes)
		{
		case DIRECTIONAL:
			myLightSources[index].isEnabled = true;
			myLightSources[index].isLocal = false;
			myLightSources[index].isSpot = false;
			memcpy(myLightSources[index].l_pos, pos, sizeof(float)* 4);
			break;

		case POINT:
			myLightSources[index].isEnabled = false;
			myLightSources[index].isLocal = true;
			myLightSources[index].isSpot = false;
			memcpy(myLightSources[index].l_pos, pos, sizeof(float)* 4);
			break;

		case SPOT:
			myLightSources[index].isEnabled = false;
			myLightSources[index].isLocal = true;
			myLightSources[index].isSpot = true;
			memcpy(myLightSources[index].l_pos, pos, sizeof(float)* 4);
			memcpy(myLightSources[index].l_spotDir, spotDir, sizeof(float)* 4);
			myLightSources[index].l_spotCutOff = spotCutOff;
			break;
		}
	}
}

void Light::draw(VSShaderLib shader) {

	GLint loc;
	float res[4];
	char s[25];

	for (int i = 0; i < MAX_LIGHTS; i++) {

		//Enabled bool
		sprintf(s, "lights[%d].isEnabled", i);
		loc = glGetUniformLocation(shader.getProgramIndex(), s);
		glUniform1i(loc, myLightSources[i].isEnabled);

		//Local bool
		sprintf(s, "lights[%d].isLocal", i);
		loc = glGetUniformLocation(shader.getProgramIndex(), s);
		glUniform1i(loc, myLightSources[i].isLocal);

		//Spot bool
		sprintf(s, "lights[%d].isSpot", i);
		loc = glGetUniformLocation(shader.getProgramIndex(), s);
		glUniform1i(loc, myLightSources[i].isSpot);

		if (myLightSources[i].isLocal) {

			//Light position
			sprintf(s, "lights[%d].l_pos", i);
			vsml->multMatrixPoint(VSMathLib::VIEW, myLightSources[i].l_pos, res);
			loc = glGetUniformLocation(shader.getProgramIndex(), s);
			glUniform4fv(loc, 1, res);

			if (myLightSources[i].isSpot) {

				//Spotlight direction
				sprintf(s, "lights[%d].l_spotDir", i);
				vsml->multMatrixPoint(VSMathLib::VIEW, myLightSources[i].l_spotDir, res);
				loc = glGetUniformLocation(shader.getProgramIndex(), s);
				glUniform4fv(loc, 1, res);

				//Spotlight cutoff
				sprintf(s, "lights[%d].l_spotCutOff", i);
				loc = glGetUniformLocation(shader.getProgramIndex(), s);
				glUniform1f(loc, myLightSources[i].l_spotCutOff);
			}
		}
		else {

			//Light position
			sprintf(s, "lights[%d].l_pos", i);
			vsml->multMatrixPoint(VSMathLib::VIEW, myLightSources[i].l_pos, res);
			loc = glGetUniformLocation(shader.getProgramIndex(), s);
			glUniform4fv(loc, 1, res);
		}
	}
}

void Light::switchLight(int index) {

	if (myLightSources[index].isEnabled) {
		myLightSources[index].isEnabled = false;
	}
	else {
		myLightSources[index].isEnabled = true;
	}
}

void Light::updateSpotPosDir(int index, float *pos, float *spotDir) {

	memcpy(myLightSources[index].l_pos, pos, sizeof(float)* 4);
	memcpy(myLightSources[index].l_spotDir, spotDir, sizeof(float)* 4);
}

File: FroggerJS/light.js
function Light() {
	
	this.lights = [];

	this.lights.push({ isEnabled: false, isLocal: false, isSpot: false, l_pos: [0,0,0,1], l_spotDir: [0,0,0,1], l_spotCutOff: 0.0})
	this.lights.push({ isEnabled: false, isLocal: false, isSpot: false, l_pos: [0,0,0,1], l_spotDir: [0,0,0,1], l_spotCutOff: 0.0})
	this.lights.push({ isEnabled: false, isLocal: false, isSpot: false, l_pos: [0,0,0,1], l_spotDir: [0,0,0,1], l_spotCutOff: 0.0})
	this.lights.push({ isEnabled: false, isLocal: false, isSpot: false, l_pos: [0,0,0,1], l_spotDir: [0,0,0,1], l_spotCutOff: 0.0})
	this.lights.push({ isEnabled: false, isLocal: false, isSpot: false, l_pos: [0,0,0,1], l_spotDir: [0,0,0,1], l_spotCutOff: 0.0})
	this.lights.push({ isEnabled: false, isLocal: false, isSpot: false, l_pos: [0,0,0,1], l_spotDir: [0,0,0,1], l_spotCutOff: 0.0})
	this.lights.push({ isEnabled: false, isLocal: false, isSpot: false, l_pos: [0,0,0,1], l_spotDir: [0,0,0,1], l_spotCutOff: 0.0})
	this.lights.push({ isEnabled: false, isLocal: false, isSpot: false, l_pos: [0,0,0,1], l_spotDir: [0,0,0,1], l_spotCutOff: 0.0})
}

Light.prototype.addLight = function(lightType, pos, spotDir, spotCutOff, index) {

	if(index < 8) {

		if(lightType == "directional") {

			this.lights[index].isEnabled = true;
			this.lights[index].isLocal = false;
			this.lights[index].isSpot = false;
			this.lights[index].l_pos = pos;

		} else if(lightType == "point") {

			this.lights[index].isEnabled = false;
			this.lights[index].isLocal = true;
			this.lights[index].isSpot = false;
			this.lights[index].l_pos = pos;

		} else if(lightType == "spot") {

			this.lights[index].isEnabled = false;
			this.lights[index].isLocal = true;
			this.lights[index].isSpot = true;
			this.lights[index].l_pos = pos;
			this.lights[index].l_spotDir = spotDir;
			this.lights[index].l_spotCutOff = spotCutOff;
		}
	}
}

Light.prototype.draw = function(shaderProgram) {

	for(var i = 0; i < 8; i++) {

		//Enabled bool
		var loc = gl.getUniformLocation(shaderProgram, String.format("lights[{0}].isEnabled", i));
		gl.uniform1i(loc, this.lights[i].isEnabled);

		//Local bool
		loc = gl.getUniformLocation(shaderProgram, String.format("lights[{0}].isLocal", i));
		gl.uniform1i(loc, this.lights[i].isLocal);

		//Spot bool
		loc = gl.getUniformLocation(shaderProgram, String.format("lights[{0}].isSpot", i));
		gl.uniform1i(loc, this.lights[i].isSpot);

		if(this.lights[i].isLocal) {

			//Light position
			loc = gl.getUniformLocation(shaderProgram, String.format("lights[{0}].l_pos", i));
			gl.uniform4fv(loc, multMatrixPoint(this.lights[i].l_pos, vMatrix));

			if(this.lights[i].isSpot) {

				//Spotlight direction
				loc = gl.getUniformLocation(shaderProgram, String.format("lights[{0}].l_spotDir", i));
				gl.uniform4fv(loc, multMatrixPoint(this.lights[i].l_spotDir, vMatrix));

				//Spotlight cutoff
				loc = gl.getUniformLocation(shaderProgram, String.format("lights[{0}].l_spotCutOff", i));
				gl.uniform1f(loc, this.lights[i].l_spotCutOff);
			}
		}
		else {

			//Light position
			loc = gl.getUniformLocation(shaderProgram, String.format("lights[{0}].l_pos", i));
			gl.uniform4fv(loc, multMatrixPoint(this.lights[i].l_pos, vMatrix));
		}
	}
}

Light.prototype.switchLight = function(index) {

	if (this.lights[index].isEnabled) {
		this.lights[index].isEnabled = false;
	}
	else {
		this.lights[index].isEnabled = true;
	}
}